README.md
# Generative Art with Flow Fields
- Framework: p5.js
- Playing with Perlin Noise

## To recreate this:
![Screenshot](./files/images/tyler-hobbs-art.jpg)


CHANGELOG.md
# Change Log (Commits)

<!--
Major.Minor.Patch:

Major:
- New Feature(s) that break API
- Stable release

Minor:
- New feature
- Feature change
- Feature removal
- Dependencies minor/major update
- Security fix

Patch:
- New content
- Update content
- Content removal
- Dependencies patch update
- Layout improvement
- Performance improvement
-->


<!-- "In this commit I ..." -->


# "Create first basic generative artworks"
##### 008 - 05.12.2021, v0.1.8


# "Add color and noise to strokes"
##### 007 - 05.12.2021, v0.1.7


# "Create basic blob of strokes"
##### 006 - 05.12.2021, v0.1.6


# "Clean up code more"
##### 005 - 05.12.2021, v0.1.5


# "Clean up code"
##### 005 - 05.12.2021, v0.1.4


# "Create groups of flow field strokes"
##### 004 - 04.12.2021, v0.1.3


# "Create single stroke following the flow field"
##### 003 - 03.12.2021, v0.1.2


# "Create basic stroke"
##### 002 - 02.12.2021, v0.1.1


# "Initial commit with basic flow field animation"
##### 001 - 02.12.2021, v0.1.0

js/particle.js
function Particle() {
  this.pos = createVector(random(width), random(height));
  this.vel = createVector(0,0);
  this.acc = createVector(0,0);
  this.maxSpeed = 1;
  this.prevPos = this.pos.copy();

  this.update = function() {
    this.vel.add(this.acc);
    this.vel.limit(this.maxSpeed);
    this.pos.add(this.vel);
    this.acc.mult(0);
  };

  this.follow = function(vectors) {
    var x = floor(this.pos.x/scl);
    var y = floor(this.pos.y/scl);
    var index = x + y * cols;
    var force = vectors[index];
    this.applyForce(force);
  }

  this.applyForce = function(force) {
    this.acc.add(force);
  }

  // edit appearance of particles
  this.show = function() {
    // stroke(0, 0.8); // take out opacity to make particles fully visible
    strokeWeight(2);
    point(this.pos.x, this.pos.y);
    // // Connect Dots
    // line(this.pos.x, this.pos.y, this.prevPos.x, this.prevPos.y);
    // this.updatePrev();
  }

  this.updatePrev = function() {
    this.prevPos.x = this.pos.x;
    this.prevPos.y = this.pos.y;
  }

  // Make particles enter back scene when leaving
  this.edges = function() {
    if (this.pos.x > width) {
      this.pos.x = 0;
      // this.updatePrev();
    };
    if (this.pos.x < 0) {
      this.pos.x = width;
      // this.updatePrev();
    };
    if (this.pos.y > height) {
      this.pos.y = 0;
      // this.updatePrev();
    };
    if (this.pos.y < 0) {
      this.pos.y = height;
      // this.updatePrev();
    };
  }
}

archive/stroke.js
// Stroke
let strokeSteps = 16;
let strokeDistance = 1;
let strokeLength = 40;
let strokeX = 200;
let strokeY = 200;

let increment = 5; //variance


// functions
function createSingleStroke(index, line) {
  let noiseX1 = noise(index) * increment;
  let noiseX2 = noise(index + 0.1) * increment;

  let posY = strokeY + index * strokeDistance;

  // beginShape();
  //   curveVertex(0, 0);
  //   curveVertex((strokeX - strokeLength) - noiseX1, posY - noiseX1);
  //   curveVertex((strokeX + strokeLength) + noiseX2, posY + noiseX2);
  //   curveVertex(400, 400);
  // endShape();


  // Draw curved line
  beginShape();
  for (let i = 0; i < line.length; i++) {
    // console.log("line", i, line[i]);
    curveVertex(line[i].x, line[i].y + index * strokeDistance);
  };
  endShape();
};

function createStrokes(noLines = 5, weight = 1, color = 0, line) {
  strokeWeight(weight);
  stroke(color)

  for (let i = 0; i < noLines/2; i++) {
    createSingleStroke(i, line)
    // avoid drawing 2 lines on top of each other for i=0 and i=1
    if (i > 0) {
      createSingleStroke(i*-1, line);
    };
  };
};

function paintStroke(line) {
  // Multiply Lines centered
  createStrokes(strokeSteps, 4, 0, line);
  createStrokes(strokeSteps, 2, 190, line);
}


js/playground/flowField.js
var dim = 400;
var inc = 0.1;
var scl = 10;

var cols, rows;

var zoff = 0;
var fr;

var particles = [];
var noParticles = 600;

var flowField;

function setup() {
  background(255);
  createCanvas(dim, dim);
  cols = floor(width/scl);
  rows = floor(height/scl);
  for (let i = 0; i < noParticles; i++) {
    particles[i] = new Particle();
  }
  flowField = new Array(rows*cols);

  // Paragraph
  fr = createP("");
}

function draw() {
  // randomSeed(10); // stop randomness from animating
  var yoff = 0;
  for (var y = 0; y < rows; y++) {
    var xoff = 0;
    for (var x = 0; x < cols; x++) {
      var index = x + y * cols;
      var angle = noise(xoff, yoff, zoff) * TWO_PI;
      var v = p5.Vector.fromAngle(angle);
      
      v.setMag(4); // INFO: how strong force (vector) is applied to particles
      flowField[index] = v;
      xoff += inc;
    }
    yoff += inc;

    zoff += 0.00005; // INFO: how fast vectors change over time
  }

stroke(0, 0.8);
  for (let i = 0; i < particles.length; i++) {
    particles[i].follow(flowField);
    particles[i].update();
    particles[i].edges();
    particles[i].show();
  }
  
  // fr.html(floor(frameRate()));
}


js/playground/vectorField.js
var dim = 400;
var inc = 0.1;
var scl = 20;

var cols, rows;

var zoff = 0;
var fr;

var particles = [];
var noParticles = 600;

var flowField;

function setup() {
  createCanvas(dim, dim);
  cols = floor(width/scl);
  rows = floor(height/scl);
  for (let i = 0; i < noParticles; i++) {
    particles[i] = new Particle();
  }
  flowField = new Array(rows*cols);

  // Paragraph
  fr = createP("");
}

function draw() {
  background(255);
  // randomSeed(10); // stop randomness from animating
  var yoff = 0;
  for (var y = 0; y < rows; y++) {
    var xoff = 0;
    for (var x = 0; x < cols; x++) {
      var index = x + y * cols;
      var angle = noise(xoff, yoff, zoff) * TWO_PI;
      var v = p5.Vector.fromAngle(angle);
      
      v.setMag(4); // INFO: how strong force (vector) is applied to particles
      flowField[index] = v;
      xoff += inc;

      // // Draw rectangle with noise values
      // fill(r);
      // rect(x * scl, y * scl, scl, scl);
      
      // // Draw vector indicator
      stroke(0, 50);
      strokeWeight(1);
      push();
      translate(x * scl, y * scl);
      rotate(v.heading());
      line(0, 0, scl, 0);
      pop();
    }
    yoff += inc;

    // zoff += 0.00005; // INFO: how fast vectors change over time
  }

  stroke(0); // stroke(0, 0) to make particles invisible
  for (let i = 0; i < particles.length; i++) {
    particles[i].follow(flowField);
    particles[i].update();
    particles[i].edges();
    particles[i].show();
  }
  
  // fr.html(floor(frameRate()));
}


js/playground/strokeGroups.js
let dim = {x: 1000, y: 1000};
let inc = 0.1;
let scl = 25;

const noStrokeLines = 6;
const noStrokes = 500;

let cols, rows;
let zoff = 0;
let flowField;
let fr;

let lineWeight = dim.x / (18*scl);

function setup() {
  createCanvas(dim.x, dim.y);

  cols = floor(width/scl);
  rows = floor(height/scl);
  flowField = new Array(rows*cols);

  noFill();
  noLoop();

  // Paragraph
  fr = createP("");
}

function draw() {
  background(255);
  
  // Create flow field
  let yoff = 0;
  for (let y = 0; y < rows; y++) {
    let xoff = 0;
    for (let x = 0; x < cols; x++) {
      let index = x + y * cols;
      let angle = noise(xoff, yoff, zoff) * TWO_PI;
      let v = p5.Vector.fromAngle(angle);
      flowField[index] = {vector: v};
      flowField[index].position = {x, y};
      xoff += inc;
      // showVectorField(x, y, v);
    }
    yoff += inc;
    // zoff += 0.00005; // INFO: how fast vectors change over time
  }

  // Draw stylized strokes
  drawStrokeGroup(80);
  drawStrokeGroup(160);
  drawStrokeGroup(240);
  
  // fr.html(floor(frameRate()));
}

function showVectorField(x, y, v) {
  // Draw vector indicator
      stroke(0, 50);
      strokeWeight(1);
      push();
      translate(x * scl, y * scl);
      rotate(v.heading());
      line(0, 0, scl, 0);
      pop();
}

function drawPoint(p, c) {
  stroke(c)
  strokeWeight(3);
  point(p.x, p.y);
}

function getFlowFieldElement(point) {
  const gridPos = {x: floor(point.x/scl), y: floor(point.y/scl)};
  return flowField.find(element => element.position.x == gridPos.x && element.position.y == gridPos.y) || undefined;
}

// Creates a number of points that follow the flow field
function createFlowLine(startPoint, noPoints) {
  const line = [startPoint];
  for (let i = 0; i < noPoints - 1; i++) {
    const point = line[i].copy();
    // console.log("point", i, point);
    // drawPoint(point, floor(255/noPoints) * i);
    const force = getFlowFieldElement(point).vector;
    const normForce = force.normalize().mult(scl);
    const nextPoint = point.add(normForce);
    // Prevents long strokes
    if (nextPoint.x < 0 || nextPoint.x > dim.x || nextPoint.y < 0 || nextPoint.y > dim.y) {
      break;
    }
    line.push(nextPoint.copy());
  };
  // drawPoint(line[line.length - 1], 200);
  // console.log("line length", line.length);
  if (line.length !== noPoints) {
    return null
  } else {
    return line;
  }
};
// Creates a number of points that follow the flow field
function createAdjacentFlowLine(line, no) {
  const adjLine = [];
  for (let i = 0; i < line.length; i++) {
    const lineDistance = 2;
    let point = line[i].copy();
    const force = getFlowFieldElement(point).vector;
    const adjForce = force.copy().setMag(lineDistance * no).rotate(-PI/2);
    point = point.add(adjForce);
    // drawPoint(point, i * 30);
    adjLine.push(point);
  };
  // console.log(adjLine);
  return adjLine;
};

function drawCurvedStroke(lines, weight, color) {
  // console.log("Random line:", line);
  strokeWeight(weight);
  stroke(color)
  
  for (let i = 0; i < lines.length; i++) {
    beginShape();
    for (let ii = 0; ii < lines[i].length; ii++) {
      const line = lines[i];
      curveVertex(line[ii].x, line[ii].y);
    }
    endShape();
  };
};

function stylizedStroke(line, color) {
  drawCurvedStroke(line, lineWeight*2.5, 0);
  drawCurvedStroke(line, lineWeight, color);
}

function drawStrokeGroup(color) {
  let counter = 0;
  let counter2 = 0; 
  while (counter < noStrokes) {
    // console.log("counter", counter);
    // Get random position in flow field and get vector
    const randomPoint = createVector(random(width), random(height));
    // Draw base line starting from random position
    const baseLine = createFlowLine(randomPoint, 4);
    // console.log("Line length", baseLine);
    // Draw more lines base on base line to create stroke
    if (baseLine) {
      // console.log("stroke OK", counter2);
      const stroke = [baseLine];
      for (let i = 1; i < noStrokeLines; i++) {
        const adjLines = createAdjacentFlowLine(baseLine, i);
        stroke.push(adjLines);
      };
      // console.log("Stroke", stroke);
      stylizedStroke(stroke, color);
      counter++
    }
    counter2++
    if (counter2 > 1500) {
        counter = 10000000;
      };
  };
}

js/sketch.js
let cols, rows;
let flowField;
let frameRateCheck;
let zoff = 0;

// Canvas
let dim = {x: 1000, y: 1000};
let sceneScale = 25;
let inc = 0.1;

// Strokes
let noStrokes = 50;
let noStrokeLines = 6;
const strokeLineDistance = 1.9;
const strokeLineNoise = 0.35;
const lineNoise = 0.9;

let lineWeight = dim.x / (18*sceneScale);

// Spots
let spotArea = 2;
let noSpotStrokes = 100;
let spotNoise = 5;

// Colors
const colorPalette1 = ["#2B3A42", "#3F5866", "#BDD3DE", "#F0F0DF", "#FF9000"]; // Kuler: machine
const colorPalette2 = ["#272942", "#565773", "#CCD0DE", "#FFFFFF", "#FF2612"]; // Kuler: machine
const colorPalette3 = ["#609BE6", "#65C6F0", "#66D5D9", "#65F0CD", "#60E69F"]; // Kuler - colorful

const outlineColor = "#000";
const colorPalette = colorPalette2;


function setup() {
  createCanvas(dim.x, dim.y);
  cols = floor(width/sceneScale);
  rows = floor(height/sceneScale);
  flowField = new Array(rows*cols);
  noFill();
  noLoop();
  frameRateCheck = createP(""); // Paragraph
}

function draw() {
  background(255);
  // Create flow field
  let yoff = 0;
  for (let y = 0; y < rows; y++) {
    let xoff = 0;
    for (let x = 0; x < cols; x++) {
      let index = x + y * cols;
      let angle = noise(xoff, yoff, zoff) * TWO_PI;
      let v = p5.Vector.fromAngle(angle);
      flowField[index] = {vector: v};
      flowField[index].position = {x, y};
      xoff += inc;
      // showVectorField(x, y, v);
    };
    yoff += inc;
    // zoff += 0.00005; // INFO: how fast vectors change over time
  };

  // // Draw blob of strokes
  // drawBlobsOfStrokes(5);

  // // Draw random single strokes
  // const noStrokeTypes = 5
  // for (let i = 0; i < noStrokeTypes; i++) {
  //   drawSingleStrokes(pickColorFromPalette());
  // };

// ARTWORK
  // Big spots 
  spotArea = 6;
  noSpotStrokes = 200;
  drawBlobsOfStrokes(200);
  // Medium spots
  spotArea = 2;
  noSpotStrokes = 50;
  spotNoise = 10;
  drawBlobsOfStrokes(50);
  // Many Single strokes
  const noStrokeTypes = 10
  for (let i = 0; i < noStrokeTypes; i++) {
    drawSingleStrokes(pickColorFromPalette());
  };


  // frameRateCheck.html(floor(frameRate()));
  console.log("Job done!");
};

// --------------------------------------------------
// Functions
// --------------------------------------------------

function drawPoint(p, c) {
  stroke(c)
  strokeWeight(10);
  point(p.x, p.y);
};

function showVectorField(x, y, v) {
  // Draw vector indicator
      stroke(0, 50);
      strokeWeight(1);
      push();
      translate(x * sceneScale, y * sceneScale);
      rotate(v.heading());
      line(0, 0, sceneScale, 0);
      pop();
};

function pickColorFromPalette() {
  const index = round(map(random(), 0, 1, 0, colorPalette.length - 1))
  const color = colorPalette[index];
  // console.log("Color index", index);
  // console.log("Color", color);
  return color;
}

function randomNoise(factor) {
  return map(noise(random()), 0, 1, 1 - factor, 1 + factor);
};

function getFlowFieldElement(point) {
  const gridPos = {x: floor(point.x/sceneScale), y: floor(point.y/sceneScale)};
  return flowField.find(element => element.position.x == gridPos.x && element.position.y == gridPos.y);
};

// Creates a number of points that follow the flow field
function createFlowLine(startPoint, noPoints) {
  const line = [startPoint];
  for (let i = 0; i < noPoints - 1; i++) {
    const point = line[i].copy();
    // console.log("point", i, point);
    // drawPoint(point, floor(255/noPoints) * i);
    const direction = getFlowFieldElement(point).vector;
    const force = direction.setMag(sceneScale);
    const nextPoint = point.add(force);
    // Prevents long strokes due to points leaving the canvas
    if (nextPoint.x < 0 || nextPoint.x > dim.x || nextPoint.y < 0 || nextPoint.y > dim.y) {
      break;
    };
    line.push(nextPoint.copy());
  };
  // drawPoint(line[line.length - 1], 200);
  // console.log("line length", line.length);
  if (line.length !== noPoints) {
    return null;
  } else {
    return line;
  };
};

// Creates a number of points that follow the flow field
function createAdjacentFlowLine(line, no) {
  const adjacentLine = [];
  for (let i = 0; i < line.length; i++) {
    let point = line[i].copy();
    const direction = getFlowFieldElement(point).vector;
    const force = direction.copy().setMag(no * strokeLineDistance * randomNoise(strokeLineNoise)).rotate(-PI/2  * randomNoise(lineNoise));
    point = point.add(force);
    // drawPoint(point, i * 30);
    adjacentLine.push(point);
  };
  // console.log(adjacentLine);
  return adjacentLine;
};

function drawCurvedStroke(lines, weight, color) {
  strokeWeight(weight);
  stroke(color);
  for (let i = 0; i < lines.length; i++) {
    beginShape();
    for (let ii = 0; ii < lines[i].length; ii++) {
      const line = lines[i];
      curveVertex(line[ii].x, line[ii].y);
    };
    endShape();
  };
};

function stylizedStroke(line, color) {
  drawCurvedStroke(line, lineWeight * 2.5, outlineColor);
  drawCurvedStroke(line, lineWeight, color);
};

function drawSingleStrokes(color) {
  let counter = 0;
  while (counter < noStrokes) {
    // console.log("counter", counter);
    const randomPoint = createVector(random(width), random(height));
    // Draw base line with 4 points starting from random position
    const baseLine = createFlowLine(randomPoint, 4);
    // console.log("Line length", baseLine);
    // Draw more lines base on base line to create stroke (if not null)
    if (baseLine) {
      const stroke = [baseLine];
      for (let i = 1; i < noStrokeLines; i++) {
        const adjacentLines = createAdjacentFlowLine(baseLine, i);
        stroke.push(adjacentLines);
      };
      // console.log("Stroke", stroke);
      stylizedStroke(stroke, color);
      counter++;
    };
  };
};

// Blob of same stroke
function drawBlobsOfStrokes(noSpots, color) {
  let isColorDefined;
  color ? isColorDefined = true : isColorDefined = false;
  // Spots
  for (let i = 0; i < noSpots; i++) {
    const spots = [];
    const randomSpot = createVector(random(width), random(height));
    const spotRadius = spotArea * sceneScale * randomNoise(spotNoise);

    // When no color defined, pick a random one from current color palette
      if (!isColorDefined) {
        color = pickColorFromPalette();
      };
    
    // console.log("random Spot", randomSpot.x, randomSpot.y);
    let counter = 0;
    let loopChecker = 0;
    while (counter < noSpotStrokes) {
      const spotPoint = createVector((randomSpot.x - spotRadius) + (2 * spotRadius * noise(random() * counter)), ((randomSpot.y - spotRadius) + (2 * spotRadius * noise(random() * counter))));
      
      // Avoid issues with line tracking in flow field
      if (spotPoint.x < 0) {
        spotPoint.x = 10;
      };
      if (spotPoint.x > width) {
        spotPoint.x = width - 10;
      };
      if (spotPoint.y < 0) {
        spotPoint.y = 10 ;
      };
      if (spotPoint.y > height) {
        spotPoint.y = height - 10;
      };

      // drawPoint(spotPoint, 200);
      // console.log("Spot Point", spotPoint.x, spotPoint.y);
      // Draw base line with 4 points starting from spotPoint handle position
      const baseLine = createFlowLine(spotPoint, 4);
      // console.log("Line length", baseLine);
      // Draw more lines base on base line to create stroke (if not null)
      if (baseLine) {
        const stroke = [baseLine];
        for (let i = 1; i < noStrokeLines; i++) {
          const adjacentLines = createAdjacentFlowLine(baseLine, i);
          stroke.push(adjacentLines);
        };
        // console.log("Stroke", stroke);
        counter++;
        spots.push(stroke);
      } else {
        loopChecker++
        console.warn("Cannot create stroke:", spotPoint.x, spotPoint.y);
      };
      // Get out of loop if there are issues
      if (loopChecker > 1000) {
        counter = noSpotStrokes
        console.error("LOOP EXIT - Strokes too close to canvas border.");
      };
    };
    // console.log("NO of Elements", counter, noSpotStrokes);
      for (let j = 0; j < spots.length; j++) {
        drawCurvedStroke(spots[j], lineWeight * 2.5, outlineColor);
      };
      for (let k = 0; k < spots.length; k++) {
        drawCurvedStroke(spots[k], lineWeight, color);
      };
      // drawPoint(randomSpot, 0);
  };
};